slyd/slyd/orm/base.py

```diff
@@ -0,0 +1,550 @@
+    def __new__(mcs, name, bases, attrs):
+        parents = [b for b in bases if isinstance(b, ModelMeta)]
+        if not parents:
+            return super(ModelMeta, mcs).__new__(mcs, name, bases, attrs)
+
+        # check if a model with the same name exists in the registry
+        if name in models:
+            raise ImproperlyConfigured(
+                u"A Model named '{}' already exists".format(name))
+
+        meta = attrs.pop('Meta', None)
+        primary_key = None
+        fields = {}
+        basic_attrs = {}
+        file_schema_attrs = {}
+
+        for attrname, value in iteritems(attrs):
+            if isinstance(value, BaseRelationship):
+                fields[attrname] = value
+            elif isinstance(value, Field):
+                if value.primary_key:
+                    if primary_key:
+                        raise ImproperlyConfigured(
+                            u"Model '{}' declared with more than one primary "
+                            u"key".format(name))
+                    primary_key = attrname
+                fields[attrname] = value
+            # move decorated marshmallow methods to the file schema
+            elif hasattr(value, '__marshmallow_tags__'):
+                file_schema_attrs[attrname] = value
+            else:
+                basic_attrs[attrname] = value
+
+        if fields and not primary_key:
+            raise ImproperlyConfigured(
+                u"Model '{}' declared with no primary key".format(name))
+
+        dependencies = {}
+        for attrname, field in iteritems(fields):
+            if attrname == primary_key:
+                dependencies[attrname] = set()
+            elif isinstance(field, DependantField):
+                dependencies[attrname] = {primary_key, field.when}
+            else:
+                dependencies[attrname] = {primary_key}
+        try:
+            ordered_fields = toposort_flatten(dependencies)
+        except ValueError as e:
+            raise ImproperlyConfigured(e.message)
+
+        cls = super(ModelMeta, mcs).__new__(mcs, name, bases, basic_attrs)
+        cls._pk_field = primary_key
+        cls._fields = fields
+        cls._ordered_fields = ordered_fields
+        cls._file_fields = file_fields = {k for k, f in iteritems(fields)
+                                          if not f.ignore_in_file}
+        cls._field_names = sorted(k for k, f in iteritems(fields)
+                                  if isinstance(f, Field))
+        cls._relationship_names = sorted(k for k, f in iteritems(fields)
+                                         if isinstance(f, BaseRelationship))
+        cls.opts = ModelOpts(meta, cls)
+        cls.collection = type(name + 'Collection', (ModelCollection,), {
+            'model': cls
+        })
+
+        for attrname, field in iteritems(fields):
+            if attrname in file_fields:
+                file_schema_attrs[attrname] = field
+            field.contribute_to_class(cls, attrname)
```

Comment:
Right now we are delegating a call to `settattr(cls, ...)` to `field` which handles descriptor construction using duck typing on `Field` and `List` (fields.py). However, I think that `setattr` should be called in the meta class as the main responsibility of the meta class is this kind of "surgery" before a class is instantiated. A perhaps better alternative is to define a method which will build a descriptor based on the `attrname`:
```
class Field(fields.Field):
    def build_descriptor(self, attrname):
        return FieldDescriptor(attrname, self)

class List(fields.List, Field):
    def build_descriptor(self, attrname):
        return ListDescriptor(attrname=attrname)
```
and then replace `field.contribute_to_class(cls, attrname)` with `setattr(cls, attrname, field.build_descriptor(attrname))`.